### ISSUE TITLE
Language Server will not start

### ISSUE THREAD

**Issue by jreitmayer** [lingua-pupuli/puppet-vscode#453]
### What Versions are you running?

OS Version: Windows 10
VSCode Version: 1.30.1
Puppet Extension Version: Not in use
Puppet Agent Version: Not in use
PDK Version (if applicable): 1.80

### What You Are Seeing?

Language Server will not start

### What is Expected?

language server starts

### How Did You Get This To Happen? (Steps to Reproduce)

just start vscode

### Output Log
```
C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `mkdir': Permission denied @ dir_s_mkdir - M:/.puppetlabs (Errno::EACCES)
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `fu_mkdir'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:208:in `block (2 levels) in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `reverse_each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `block in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:363:in `create_ancestors'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:348:in `block in initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:150:in `do_initialize_settings_for_run_mode'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:142:in `initialize_settings'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/lib/puppet_languageserver.rb:143:in `init_puppet'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/puppet-languageserver:10:in `<main>'
[Info  - 2:30:01 PM] Connection to server got closed. Server will restart.
C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `mkdir': Permission denied @ dir_s_mkdir - M:/.puppetlabs (Errno::EACCES)
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `fu_mkdir'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:208:in `block (2 levels) in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `reverse_each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `block in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:363:in `create_ancestors'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:348:in `block in initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:150:in `do_initialize_settings_for_run_mode'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:142:in `initialize_settings'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/lib/puppet_languageserver.rb:143:in `init_puppet'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/puppet-languageserver:10:in `<main>'
[Info  - 2:30:04 PM] Connection to server got closed. Server will restart.
C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `mkdir': Permission denied @ dir_s_mkdir - M:/.puppetlabs (Errno::EACCES)
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `fu_mkdir'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:208:in `block (2 levels) in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `reverse_each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `block in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:363:in `create_ancestors'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:348:in `block in initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:150:in `do_initialize_settings_for_run_mode'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:142:in `initialize_settings'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/lib/puppet_languageserver.rb:143:in `init_puppet'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/puppet-languageserver:10:in `<main>'
[Info  - 2:30:07 PM] Connection to server got closed. Server will restart.
C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `mkdir': Permission denied @ dir_s_mkdir - M:/.puppetlabs (Errno::EACCES)
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `fu_mkdir'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:208:in `block (2 levels) in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `reverse_each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `block in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:363:in `create_ancestors'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:348:in `block in initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:150:in `do_initialize_settings_for_run_mode'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:142:in `initialize_settings'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/lib/puppet_languageserver.rb:143:in `init_puppet'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/puppet-languageserver:10:in `<main>'
[Info  - 2:30:10 PM] Connection to server got closed. Server will restart.
C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `mkdir': Permission denied @ dir_s_mkdir - M:/.puppetlabs (Errno::EACCES)
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:230:in `fu_mkdir'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:208:in `block (2 levels) in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `reverse_each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:206:in `block in mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/private/ruby/2.4.4/lib/ruby/2.4.0/fileutils.rb:191:in `mkdir_p'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:363:in `create_ancestors'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:348:in `block in initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `each'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet/settings.rb:347:in `initialize_app_defaults'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:150:in `do_initialize_settings_for_run_mode'
	from C:/Program Files/Puppet Labs/DevelopmentKit/share/cache/ruby/2.4.0/gems/puppet-6.0.4-x64-mingw32/lib/puppet.rb:142:in `initialize_settings'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/lib/puppet_languageserver.rb:143:in `init_puppet'
	from c:/Users/jeffr/.vscode/extensions/jpogran.puppet-vscode-0.15.0/vendor/languageserver/puppet-languageserver:10:in `<main>'
[Error - 2:30:13 PM] Connection to server got closed. Server will not be restarted.
```

**Comment by jpogran** [comment]
Hi @jreitmayer, at first glance the issue appears to be that Puppet thinks your config directory is `M:/.puppetlabs`. This likely is happening because your user profile is a roaming profile or your configuration is pointing to the M drive as primary. Can you confirm?

**Comment by jreitmayer** [comment]
That is correct, can I change that?

Jeff Reitmayer  | Lead Systems Administrator
Information Technology | Division of Student Affairs
1259 TAMU | College Station, TX 77843-1259

ph: 979.458.9224 | jreitmayer@tamu.edu<mailto:jreitmayer@tamu.edu> | doit.tamu.edu<http://doit.tamu.edu/>
- - - - - - - - - - - - - - - - - - - - - - - -
TEXAS A&M UNIVERSITY

**Comment by jpogran** [comment]
So the directories are explained here: https://puppet.com/docs/puppet/3.8/dirs_confdir.html#windows-systems, the relevant line being the last in that table referring to 'non-administrator'. Since the confdir is special, it can't be set in the puppet.conf as it needs to be known before things start getting initialized. The way to set normally it is by passing the `--confdir` parameter to Puppet. 

However we aren't running Puppet from the commandline, we're running it inside the Puppet Language Server, which is started by the Puppet VSCode Extension. The Puppet Language Server has the command line option `--puppet-settings`, which would allow you to pass in `--confdir`, but you the user don't have access to that from inside the Puppet Extension (the side you see) yet.

In short, the bad news is we don't have a way for you to change how our extension passes the Puppet specific settings yet. The good news is we can fix it, I'll add this to our board.

**Comment by jreitmayer** [comment]
Thanks. Sorry for all the trouble.

**Comment by glennsarti** [comment]
@jreitmayer  PR is up to add the setting you need.

**Comment by jreitmayer** [comment]
I'm sorry? PR? @glennsarti

**Comment by glennsarti** [comment]
@jreitmayer Sorry .. PR = Pull Request.

I've created #462 which will add the setting you need so you can add in the `confdir`

**Comment by jreitmayer** [comment]
@glennsarti how do i update the puppet plugin then? Forgive me, I am new to all this.

**Comment by glennsarti** [comment]
Sure thing 👍

The short version; Your problem is being worked on.  As a courtesy I tell people where I'm at with fixing it so they know what's going on

long version;
I have made the code changes we need to add the setting.  I've raised a pull request (PR) so @jpogran can review my code, and if he's happy it'll take my changes and apply them.  Then on the next release cycle (usually end of month) we'll release an updated Extension (I think 0.16.0 is the next one) and then you can install it (although it updates automatically anyways so...)

Then we can rejoice for your problem is solved!!

**Comment by jreitmayer** [comment]
Awesome boss. Great job. Thanks for the work.
